tests/test_kernel.py

```diff
@@ -597,6 +598,43 @@ def test_control_thread_priority():
     assert control_dates[-1] <= shell_dates[0]
 
 
+def test_sequential_control_messages():
+    with new_kernel() as kc:
+        msg_id = kc.execute("import time")
+        get_reply(kc, msg_id)
+
+        # Send multiple messages on the control channel.
+        # Using execute messages to vary duration.
+        sleeps = [0.6, 0.3, 0.1]
+
+        # Prepare messages
+        msgs = [
+            kc.session.msg("execute_request", {"code": f"time.sleep({sleep})"}) for sleep in sleeps
+        ]
+        msg_ids = [msg["header"]["msg_id"] for msg in msgs]
+
+        # Submit messages
+        for msg in msgs:
+            kc.control_channel.send(msg)
+
+        # Get replies
+        replies = [get_reply(kc, msg_id, channel="control") for msg_id in msg_ids]
+
+        # Check messages are processed in order, one at a time, and of a sensible duration.
+        previous_end = None
+        for reply, sleep in zip(replies, sleeps):
+            start = datetime.fromisoformat(reply["metadata"]["started"])
+            end = reply["header"]["date"]
```

Comment:
This is interesting, in a bad way. The type of `end` differs depending on how the tests are run and is either a `datetime` or a `string`. Using
```
pytest -v tests/test_kernel.py -k control
```
they are `datetime`s (which is what I was expecting), and using
```
pytest -v tests -k control
```
they are strings. So there is something different happening depending on how the tests are invoked.

Reply:
I've worked out why the type of the header's date is different depending on how the tests are run. The cause is this line in `ipyparallel`:
https://github.com/ipython/ipyparallel/blob/5e13f5664d27b60338542c438b7d8d144546dfce/ipyparallel/client/client.py#L58
which monkey patches `jupyter_client.session.extract_dates` to be a no-op. If you are specific about which test to run then this line in `ipyparallel` is not run, but if you scan all the tests then it is run and datetimes are not deserialized from strings to datetimes.

Reply:
That looks like a bug in `ipyparallel`, mind reporting it there?

Reply:
It turns out that I cannot reproduce this problem locally with the latest `pytest-asyncio 0.23.5`, so for the moment I've just rebased this PR against `main` and will monitor it.

Reply:
Sounds good